metabase/metabase
# Refactor PivotTable Visualization into a functional typescript component

resolves https://github.com/metabase/metabase/issues/27649

## Why Refactor this component?

With the upcoming work to [make pivot table columns user-resizable](https://github.com/metabase/metabase/issues/27597), we'll need to introduce internal state to this component which will cause a large number of re-renders. To keep this performant, we'd already need to do a good bit of refactoring to memoize a bunch of calculations.  Access to the `useMemo` hook will be highly valuable.

In addition, this component is pretty big, and includes several subcomponents and calculation functions, and will be easier to understand when those are extracted.

Also, we want more stuff in typescript, and I think the work typing the internal settings for pivot tables here is really helpful in making the logic more comprehensible.

## Summary

There should be no changes to the visualization's behavior, it should just now be a functional component and properly typed.

## Testing Steps

open a big gnarly pivot table like this one:
```
http://localhost:3000/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoxLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOjIsImFnZ3JlZ2F0aW9uIjpbWyJzdW0iLFsiZmllbGQiLDEzLG51bGxdXV0sImJyZWFrb3V0IjpbWyJmaWVsZCIsMTQseyJ0ZW1wb3JhbC11bml0IjoicXVhcnRlciJ9XSxbImZpZWxkIiw4LHsic291cmNlLWZpZWxkIjo5fV0sWyJmaWVsZCIsMTIseyJiaW5uaW5nIjp7InN0cmF0ZWd5IjoiZGVmYXVsdCJ9fV0sWyJmaWVsZCIsMTcseyJiaW5uaW5nIjp7InN0cmF0ZWd5IjoiZGVmYXVsdCJ9fV1dfSwidHlwZSI6InF1ZXJ5In0sImRpc3BsYXkiOiJwaXZvdCIsImRpc3BsYXlJc0xvY2tlZCI6dHJ1ZSwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnsicGl2b3RfdGFibGUuY29sdW1uX3NwbGl0Ijp7InJvd3MiOltbImZpZWxkIiwxNCx7InRlbXBvcmFsLXVuaXQiOiJxdWFydGVyIn1dLFsiZmllbGQiLDEyLHsiYmlubmluZyI6eyJzdHJhdGVneSI6Im51bS1iaW5zIiwibWluLXZhbHVlIjowLCJtYXgtdmFsdWUiOjEwMCwibnVtLWJpbnMiOjgsImJpbi13aWR0aCI6MTIuNX19XSxbImZpZWxkIiwxNyx7ImJpbm5pbmciOnsic3RyYXRlZ3kiOiJudW0tYmlucyIsIm1pbi12YWx1ZSI6MCwibWF4LXZhbHVlIjo3MCwibnVtLWJpbnMiOjgsImJpbi13aWR0aCI6MTB9fV1dLCJjb2x1bW5zIjpbWyJmaWVsZCIsOCx7InNvdXJjZS1maWVsZCI6OX1dXSwidmFsdWVzIjpbWyJhZ2dyZWdhdGlvbiIsMF1dfX19
```
- see that the pivot table displays properly
- see that click actions work properly
- see that collapsing/expanding works properly
- open the viz settings and change stuff and see that it updates correctly
- see that performance isn't any worse than on `master` (which is pretty bad, scrollsync is slow)
